primitives/src/merkle_tree/internal.rs

```diff
@@ -544,73 +544,111 @@ where
     }
 
     /// Update the element at the given index.
-    pub(crate) fn update_internal<H, Arity>(
+    /// * `returns` - `Err()` if any error happens internally. `Ok(delta,
+    ///   result)`, `delta` represents the changes to the overall number of
+    ///   leaves of the tree, `result` contains the original lookup information
+    ///   at the given location.
+    pub(crate) fn update_with_internal<H, Arity, F>(
         &mut self,
         height: usize,
         pos: impl Borrow<I>,
         traversal_path: &[usize],
-        elem: impl Borrow<E>,
-    ) -> Result<LookupResult<E, (), ()>, PrimitivesError>
+        f: F,
+    ) -> Result<(i64, LookupResult<E, (), ()>), PrimitivesError>
     where
         H: DigestAlgorithm<E, I, T>,
         Arity: Unsigned,
+        F: FnOnce(Option<&E>) -> Option<E>,
     {
         let pos = pos.borrow();
-        let elem = elem.borrow();
         match self {
             MerkleNode::Leaf {
                 elem: node_elem,
                 value,
                 pos,
             } => {
-                let ret = ark_std::mem::replace(node_elem, elem.clone());
-                *value = H::digest_leaf(pos, elem)?;
-                Ok(LookupResult::Ok(ret, ()))
+                let result = LookupResult::Ok(node_elem.clone(), ());
+                match f(Some(node_elem)) {
+                    Some(elem) => {
+                        *value = H::digest_leaf(pos, &elem)?;
+                        *node_elem = elem;
+                        Ok((0i64, result))
+                    },
+                    None => {
+                        *self = MerkleNode::Empty;
+                        Ok((-1i64, result))
+                    },
+                }
             },
             MerkleNode::Branch { value, children } => {
-                let res = (*children[traversal_path[height - 1]]).update_internal::<H, Arity>(
+                let branch = traversal_path[height - 1];
+                let result = children[branch].update_with_internal::<H, Arity, _>(
                     height - 1,
                     pos,
                     traversal_path,
-                    elem,
+                    f,
                 )?;
-                // If the branch containing the update was not in memory, the update failed and
-                // nothing was changed, so we can short-circuit without recomputing this node's
-                // value.
-                if res == LookupResult::NotInMemory {
-                    return Ok(res);
+                if matches!(*children[branch], MerkleNode::ForgettenSubtree { .. }) {
+                    // If the branch containing the update was not in memory,
+                    // the update failed and nothing was
+                    // changed, so we can short-circuit without recomputing this
+                    // node's value.
```

Comment:
My confusion: if something is "forgotten" or "not in memory" is that because it was removed by the user? Or is this element still in the tree but it's been swapped to cache? My guess is the former. In that case, maybe use "deleted/removed by user" or something.

Reply:
Change to "forgotten by user".
The element remains in the tree (for commitment), but, like frontier, we do not track it in this MT structure.

Reply:
Now I'm even more confused. 🤪 Let C be the commitment. I remove a node but C does not change, correct? Then I serialize and deserialize, so "forgotten" nodes are now truly gone. Then I expect the commitment to change to something different from C.  Shouldn't the commitment be consistent across serialization?

Reply:
`forget` is something that you remove a not from memory, but not from the tree.
(De)serialization is w.r.t the whole tree, information won't get lost

Reply:
I think perhaps I've overlooked some code comments / rusdoc somewhere that explains this concept of "forget". Apologies and please point me to the right place.

My understanding so far:
- User may delete an item from the tree. Under the hood, such an item is "forgotten", but it's hash still contributes to the merkle root so that the root does not change even after the deletion.
- (de)serialization must remember these "forgotten" items so that the tree can be rebuilt to the same merkle root.

This seems counterintuitive to me. I expect a MT with leaves [A,B,C] to have a unique, deterministic root, independent of whether that MT was built directly from [A,B,C] vs [A,B,C,D] + delete D.